testsuite isolation is broken

* Pip version: 10.0.0b2-17-gc76452ad
* Python version: Python 3.6
* Operating system: Arch Linux

### Description:

This is the output I get when running `pip check`:
```
picard 2.0.0.dev5 requires PyQt5, which is not installed.
````

Now if I try to run part of the testsuite, e.g. the `test_freeze_user` test:
```python
py create: /home/bpierre/progs/src/pip/.tox/py
py installdeps: -r/home/bpierre/progs/src/pip/dev-requirements.txt
py develop-inst: /home/bpierre/progs/src/pip
py installed: apipkg==1.4,coverage==4.5.1,execnet==1.5.0,freezegun==0.3.10,mock==2.0.0,pbr==4.0.1,pretend==1.0.8,py==1.5.3,pytest==3.2.5,pytest-catchlog==1.2.2,pytest-cov==2.5.1,pytest-forked==0.2,pytest-rerunfailures==4.0,pytest-timeout==1.2.1,pytest-xdist==1.22.2,python-dateutil==2.7.2,PyYAML==3.12,scripttest==1.3,six==1.11.0,virtualenv==15.3.0.dev0
py runtests: PYTHONHASHSEED='900407666'
py runtests: commands[0] | py.test --timeout 300 -k test_freeze_user
============================= test session starts ==============================
platform linux -- Python 3.6.4, pytest-3.2.5, py-1.5.3, pluggy-0.4.0
rootdir: /home/bpierre/progs/src/pip, inifile: setup.cfg
plugins: xdist-1.22.2, timeout-1.2.1, rerunfailures-4.0, forked-0.2, cov-2.5.1, catchlog-1.2.2
timeout: 300.0s method: signal
collected 817 items

tests/functional/test_freeze.py F
=========================== short test summary info ============================
FAIL tests/functional/test_freeze.py::test_freeze_user

=================================== FAILURES ===================================
_______________________________ test_freeze_user _______________________________

script = <tests.lib.PipTestEnvironment object at 0x1471e39bb7f0>
virtualenv = <VirtualEnvironment /tmp/pytest-of-bpierre/pytest-13/test_freeze_user0/workspace/venv>
data = <tests.lib.TestData object at 0x1471e39bba58>

    @pytest.mark.network
    def test_freeze_user(script, virtualenv, data):
        """
        Testing freeze with --user, first we have to install some stuff.
        """
        script.pip('download', 'setuptools', 'wheel', '-d', data.packages)
        virtualenv.system_site_packages = True
        script.pip_install_local('--find-links', data.find_links,
>                                '--user', 'simple==2.0')

tests/functional/test_freeze.py:581: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/lib/__init__.py:389: in pip_install_local
    *args, **kwargs
tests/lib/__init__.py:383: in pip
    return self.run(exe, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.lib.PipTestEnvironment object at 0x1471e39bb7f0>
args = ('pip', 'install', '--no-index', '--find-links', 'file:///home/bpierre/progs/src/pip/tests/data/packages', '--find-links', ...)
kw = {}
cwd = Path('/tmp/pytest-of-bpierre/pytest-13/test_freeze_user0/workspace/scratch')
run_from = None

    def run(self, *args, **kw):
        if self.verbose:
            print('>> running %s %s' % (args, kw))
        cwd = kw.pop('cwd', None)
        run_from = kw.pop('run_from', None)
        assert not cwd or not run_from, "Don't use run_from; it's going away"
        cwd = cwd or run_from or self.cwd
        if sys.platform == 'win32':
            # Partial fix for ScriptTest.run using `shell=True` on Windows.
            args = [str(a).replace('^', '^^').replace('&', '^&') for a in args]
        return TestPipResult(
>           super(PipTestEnvironment, self).run(cwd=cwd, *args, **kw),
            verbose=self.verbose,
        )
E       AssertionError: stderr output not expected

tests/lib/__init__.py:364: AssertionError
----------------------------- Captured stdout call -----------------------------
Script result: pip install --no-index --find-links file:///home/bpierre/progs/src/pip/tests/data/packages --find-links file:///tmp/pytest-of-bpierre/pytest-13/test_freeze_user0/data/packages --user simple==2.0
-- stderr: --------------------
picard 2.0.0.dev5 requires PyQt5, which is not installed.

-- stdout: --------------------
Looking in links: file:///home/bpierre/progs/src/pip/tests/data/packages, file:///tmp/pytest-of-bpierre/pytest-13/test_freeze_user0/data/packages
Collecting simple==2.0
Building wheels for collected packages: simple
  Running setup.py bdist_wheel for simple: started
  Running setup.py bdist_wheel for simple: finished with status 'done'
  Stored in directory: /tmp/pytest-of-bpierre/pytest-13/test_freeze_user0/home/.cache/pip/wheels/fd/25/62/9fd247f390d3494522e59182e5a612f3fb69fb1150ee8b9412
Successfully built simple
Installing collected packages: simple
Successfully installed simple-2.0

-- created: -------------------
  venv/user/lib/python3.6/site-packages/simple
                      simple-2.0.dist-info
                          INSTALLER  (4 bytes)
                          METADATA  (168 bytes)
                          RECORD  (492 bytes)
                          WHEEL  (93 bytes)
                          top_level.txt  (7 bytes)
                      simple/__init__.py  (7 bytes)
-- updated: -------------------
  venv/user/lib/python3.6/site-packages
============================= 816 tests deselected =============================
=================== 1 failed, 816 deselected in 8.12 seconds ===================
ERROR: InvocationError: '/home/bpierre/progs/src/pip/.tox/py/bin/py.test --timeout 300 -k test_freeze_user'
___________________________________ summary ____________________________________
ERROR:   py: commands failed

```

The problem is the use of `virtualenv.system_site_packages = True` (so `--user` can be used).

Tentative patch:
```diff
 tests/conftest.py | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git i/tests/conftest.py w/tests/conftest.py
index e58a0c2b..48f315b1 100644
--- i/tests/conftest.py
+++ w/tests/conftest.py
@@ -157,6 +157,14 @@ def virtualenv_template(tmpdir_factory):
         pip_source_dir=pip_src,
         relocatable=True,
     )
+    site_py = venv.lib / 'site.py'
+    with open(site_py) as fp:
+        site_contents = fp.read()
+    site_pattern = '\ndef virtual_addsitepackages(known_paths):\n'
+    site_contents = site_contents.replace(site_pattern, site_pattern +
+                                          '    return known_paths\n')
+    with open(site_py, 'w') as fp:
+        fp.write(site_contents)
     if sys.platform == 'win32':
         # Work around setuptools' easy_install.exe
         # not working properly after relocation.
```

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -157,6 +157,18 @@ def virtualenv_template(tmpdir_factory):
         pip_source_dir=pip_src,
         relocatable=True,
     )
+    # Ensure `virtualenv.system_site_packages = True` (needed for testing
+    # `--user`) does not result in adding the real site-packages' directory
+    # to `sys.path`.
+    site_py = venv.lib / 'site.py'
+    with open(site_py) as fp:
+        site_contents = fp.read()
+    site_pattern = '\ndef virtual_addsitepackages(known_paths):\n'
+    assert site_pattern in site_contents
+    site_contents = site_contents.replace(site_pattern, site_pattern +
+                                          '    return known_paths\n')
+    with open(site_py, 'w') as fp:
+        fp.write(site_contents)
     if sys.platform == 'win32':
         # Work around setuptools' easy_install.exe
         # not working properly after relocation.

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -157,16 +157,37 @@ def virtualenv_template(tmpdir_factory):
         pip_source_dir=pip_src,
         relocatable=True,
     )
-    # Ensure `virtualenv.system_site_packages = True` (needed for testing
-    # `--user`) does not result in adding the real site-packages' directory
-    # to `sys.path`.
+    # Fix `site.py`.
     site_py = venv.lib / 'site.py'
     with open(site_py) as fp:
         site_contents = fp.read()
-    site_pattern = '\ndef virtual_addsitepackages(known_paths):\n'
-    assert site_pattern in site_contents
-    site_contents = site_contents.replace(site_pattern, site_pattern +
-                                          '    return known_paths\n')
+    for pattern, replace in (
+        (
+            # Ensure `virtualenv.system_site_packages = True` (needed
+            # for testing `--user`) does not result in adding the real
+            # site-packages' directory to `sys.path`.
+            (
+                '\ndef virtual_addsitepackages(known_paths):\n'
+            ),
+            (
+                '\ndef virtual_addsitepackages(known_paths):\n'
+                '    return known_paths\n'
+            ),
+        ),
+        (
+            # Fix sites ordering: user site must be added before system site.
+            (
+                '\n    paths_in_sys = addsitepackages(paths_in_sys)'
+                '\n    paths_in_sys = addusersitepackages(paths_in_sys)\n'
+            ),
+            (
+                '\n    paths_in_sys = addusersitepackages(paths_in_sys)'
+                '\n    paths_in_sys = addsitepackages(paths_in_sys)\n'
+            ),
+        ),
+    ):
+        assert pattern in site_contents
+        site_contents = site_contents.replace(pattern, replace)
     with open(site_py, 'w') as fp:
         fp.write(site_contents)
     if sys.platform == 'win32':

